infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java

```diff
@@ -0,0 +1,216 @@
+
+    final private StreamLog streamLog;
+
+    final private BlockingQueue<BatchWriterOperation> operationsQueue;
+
+    private ExecutorService processorService = Executors
+            .newSingleThreadExecutor(new ThreadFactoryBuilder()
+                    .setDaemon(false)
+                    .setNameFormat("LogUnit-BatchProcessor-%d")
+                    .build());
+
+    /**
+     * The sealEpoch is the epoch up to which all operations have been sealed. Any
+     * BatchWriterOperation arriving after the sealEpoch with an epoch less than the sealEpoch
+     * is completed exceptionally with a WrongEpochException.
+     * This is persisted in the ServerContext by the LogUnitServer to withstand restarts.
+     */
+    private long sealEpoch;
+
+    /**
+     * Returns a new BatchProcessor for a stream log.
+     *
+     * @param streamLog      stream log for writes (can be in memory or file)
+     * @param sealEpoch All operations stamped with epoch less than the epochWaterMark are
+     *                       discarded.
+     * @param streamLog the backing log (can be in memory or file)
+     * @param sync    If true, the batch writer will sync writes to secondary storage
+     */
+    public BatchProcessor(StreamLog streamLog, long sealEpoch, boolean sync) {
+        this.sealEpoch = sealEpoch;
+        this.sync = sync;
+        this.streamLog = streamLog;
+        operationsQueue = new LinkedBlockingQueue<>();
+        processorService.submit(this::processor);
+    }
+
+    /**
+     * Add a task to the processor.
+     * @param type The request type
+     * @param msg The request message
+     * @return returns a future result for the request, if it expects one
+     */
+    public <T> CompletableFuture <T> addTask(@Nonnull Type type, @Nonnull CorfuPayloadMsg msg) {
+        BatchWriterOperation<T> operation = new BatchWriterOperation<>(type, msg);
+        operationsQueue.add(operation);
+        return operation.getFutureResult();
+    }
+
+    private void processor() {
+
+        if (!sync) {
+            log.warn("batchWriteProcessor: writes configured to not sync with secondary storage");
+        }
+
+        try {
+            BatchWriterOperation lastOp = null;
+            int processed = 0;
+            List<BatchWriterOperation> res = new LinkedList<>();
+
+            while (true) {
+                BatchWriterOperation currOp;
+
+                if (lastOp == null) {
+                    currOp = operationsQueue.take();
+                } else {
+                    currOp = operationsQueue.poll();
+
+                    if (currOp == null || processed == BATCH_SIZE
+                            || currOp == BatchWriterOperation.SHUTDOWN) {
+                        streamLog.sync(sync);
+                        log.trace("Completed {} operations", processed);
+
+                        for (BatchWriterOperation operation : res) {
+                            if (!operation.getFutureResult().isCompletedExceptionally()
```

Comment:
I think we don't need to check `isCompletedExceptionally` or `isCancelled()` here. If the future is already completed exceptionally  or cancelled, completing it with any value will not overwrite the exception (i.e. the exception will still be caught in `thenRun()`, `.exceptionally()` or `get()`, etc)

Reply:
At this point, the uncompleted tasks need to be completed with their corresponding results. These are the tasks that have succeeded. L175 exceptionally completes the failed tasks. In the future, we will need to fail the whole batch if any write task failed within the batch.

Reply:
> I think we don't need to check `isCompletedExceptionally` or `isCancelled()` here. If the future is already completed exceptionally or cancelled, completing it with any value will not overwrite the exception (i.e. the exception will still be caught in `thenRun()`, `.exceptionally()` or `get()`, etc)

I think it is a kind of defensive programming and makes the code more readable.

Reply:
Alternatively, we can just call `CompletableFuture::isDone()` , but I think in this context, its more readable to check for a failure.